TrafficCapture/trafficReplayer/src/main/java/org/opensearch/migrations/replay/ReplayEngine.java

```diff
@@ -8,58 +9,92 @@
 
-    public ReplayEngine(Consumer<RequestResponsePacketPair> fullDataHandler) {
+    public ReplayEngine(Consumer<HttpMessageAndTimestamp> requestReceivedHandler,
+                        Consumer<RequestResponsePacketPair> fullDataHandler) {
         liveStreams = new HashMap<>();
+        this.requestHandler = requestReceivedHandler;
         this.fullDataHandler = fullDataHandler;
     }
 
     @Override
     public void accept(String id, TrafficObservation observation) {
-        boolean updateFirstResponseTimestamp = false;
-        RequestResponsePacketPair runningList = null;
         var pbts = observation.getTs();
         var timestamp = Instant.ofEpochSecond(pbts.getSeconds(), pbts.getNanos());
         if (observation.hasEndOfMessageIndicator()) {
-            publishAndClear(id);
+            handleEndOfMessage(id, getAccumulationForFirstRequestObservation(id));
         } else if (observation.hasRead()) {
-            runningList = liveStreams.putIfAbsent(id, new RequestResponsePacketPair());
-            if (runningList == null) {
-                runningList = liveStreams.get(id);
-                // TODO - eliminate the byte[] and use the underlying nio buffer
-                runningList.addRequestData(timestamp, observation.getRead().getData().toByteArray());
-            }
+            var runningList = getAccumulationForFirstRequestObservation(id).rrPair;
+            // TODO - eliminate the byte[] and use the underlying nio buffer
+            runningList.addRequestData(timestamp, observation.getRead().getData().toByteArray());
         } else if (observation.hasReadSegment()) {
             throw new RuntimeException("Not implemented yet.");
         } else if (observation.hasWrite()) {
-            updateFirstResponseTimestamp = true;
-            runningList = liveStreams.get(id);
+            var runningList = liveStreams.get(id).rrPair;
             if (runningList == null) {
                 throw new RuntimeException("Apparent out of order exception - " +
                         "found a purported write to a socket before a read!");
             }
             runningList.addResponseData(timestamp, observation.getWrite().toByteArray());
         } else if (observation.hasWriteSegment()) {
-            updateFirstResponseTimestamp = true;
             throw new RuntimeException("Not implemented yet.");
         } else if (observation.hasRequestReleasedDownstream()) {
-            updateFirstResponseTimestamp = true;
+
         }
-        if (updateFirstResponseTimestamp && runningList != null) {
-            if (runningList.firstTimeStampForResponse == null) {
-                runningList.firstTimeStampForResponse = timestamp;
-            }
+    }
+
+    private Accumulation getAccumulationForFirstRequestObservation(String id) {
+        var accum = liveStreams.computeIfAbsent(id, k -> new Accumulation());
+        if (accum.state == State.REQUEST_SENT) {
+            handleEndOfMessage(id, accum);
```

Comment:
Here's the new comments...
```
    // This function manages the lifecycles of the objects in the liveStreams map.
    // It will create a new Accumulation object when the first read of a request is
    // discovered, which may be for the very first observation of a TrafficStream
    // or it could be for the first read observation after the last response was
    // received.  For the latter case, this will close out the old accumulation,
    // recycling it from the liveStream map and into the response callback (by
    // invoking the callback).
    private Accumulation getAccumulationForFirstRequestObservation(String id) {
        var accum = liveStreams.computeIfAbsent(id, k -> new Accumulation());
        // If this was brand new or if we've already closed the item out and pushed
        // the state to RESPONSE_SENT, we don't need to care about triggering the
        // callback.  We only need to worry about this if we have yet to send the
        // RESPONSE.  Notice that handleEndOfMessage will bump the state itself
        // on the (soon to be recycled) accum object.
```